### PR TITLE
Major Fixes, Changes, and Improvements

### PR DESCRIPTION
**Pull request by WillsterJohnson**
Hi Alterian CX,
I came across this repo via you're ad on uk.indeed for your Bristol branch, and decided *what the hey,* I'll have a crack at solving these tests.

What I found in the repo I was a little disappointed by. I had expected this to be a great challenge to weed out the weak, but I found the challenges to be quite easy, even if the `*.spec.ts` files were a little off from the assigned requirements.

So I set about refactoring the tests themselves and, here we are.

Changes:
- All 3 challenges given a clear requirement
- All 3 challenges given full and detailed testing
- Markdown files modified for consistency and official guildline compliance
- Gitignore and Package modified to only include necessary info

If you wish to contact and/or find out more about me, my github profile has contact info and my resume is [on my website](https://willsterjohnson.com/resume/resume)

Happy coding!

### CODE REVIEW

**Comment by ct5845** [comment]
Hi,

Thanks for having a go, will close as we don't want any submissions open publicly (email in as per instructions).

**Comment by WillsterJohnson** [comment]
@ct5845 hi,
This isn't my submission, I have that separate.
This is a refactoring of the *challenges themselves*, not my attempt at the challenges.

My changes are listed above and I would be grateful if you could read (some of) the diff as I believe the changes in this PR make the challenges a much better test of a developer's ability as well as better reflect my own experience with being given requirements for a portion of a project.

Thanks

File: src/02-taxi-cab/manhattan.ts [hunk 1]
```diff
@@ -0,0 +1,8 @@
+/**
+ * Requirements:
+ * - A function `manhattan` which computes the manhattan distance between two points
+ */
+
+export const manhattan = (x1: number, y1: number, x2: number, y2: number) => {
```
Comment (ct5845) [inline src/02-taxi-cab/manhattan.ts:6]: would prefer a concrete definition of a point.

Comment (ct5845) [inline src/02-taxi-cab/manhattan.ts:6]: does 'manhattan' adequately describe the purpose of this function?

**Comment by ct5845** [comment]
Hi,

Yup read them all thanks, whilst perfectly valid in their own right, the tasks 'as refactored' don't actually reflect what we're expecting to see from individuals. 

e.g. 
```
  - `getBalance()`
    - should have a starting balance of 0
....    
  - Stick to Typescript and do not use 3rd party resources.
```
   
Have you stopped to consider this wasn't a requirement in the test for a reason? That we may want to see *how* people approach the problem, vs your amendments that actually provide more detail/more steps on how to accomplish the task. Also why not use 3rd party resources? If there's a library out there that completes a solution without you needing to write a bit of code, maybe that's a more elegant solution? (It's certainly an interesting one).

You also made probably the most critical mistake a developer can make on that first test. You removed the one/only real goal (to print a statement), so the one test that was in there, to test whether a statement would come out as expected has now disappeared. To the theoretical end-user who's requested a solution, regardless of how elegant your solution maybe, you've failed to the solve their need.

Also your "solutions" *are* committed (unless they're only partial solutions, at which point I'd dare  say there's too much in each of them ) : )

**Comment by WillsterJohnson** [comment]
Hi,

Thank you for reviewing and responding.

> Also why not use 3rd party resources?  If there's a library out there that completes a solution without you needing to write a bit of code, maybe that's a more elegant solution? (It's certainly an interesting one).

You are right that this is questionable to some degree, however its my opinion that the fewer 3rd party resources a project or snippet depends on, the more transferable and usable it becomes (of course when getting to larger scales such as a DB interface unreasonable to ask for an empty `node_modules`). I find that writing code first hand enforces good practice and more efficient packages. 
 
> You removed the one/only real goal (to print a statement), so the one test that was in there, to test whether a statement would come out as expected has now disappeared. 

If you're referring to task 01, I think this may be down to simple miscommunication. As the repo is currently, the valid solution (being one which passes the test in `account.spec.ts`) is already present in the [`account.ts`](https://github.com/AlterianCX/ui-katas/blob/master/src/01-banking/account.ts) file, meaning to pass the `account.spec.ts` no changes were necessary.
 I had assumed that this was either 
- a rouse to throw one off the scent
  In much the way that old/pre-existing code can be misleading when it comes to finding the optimal solution, I thought this may be a simulation of the same thing. Sure it *passes*, but it may not comply to the spec (eg the `Account` class)
- a piece of code left in accidentally
  Nobody's perfect, and I'll be the first to admit to making mistakes. I thought it possible that the solution in `account.ts` had been left in accidentally.
Hence I wrote up an Account class along with some interfaces and manually tested it, then worked backwards into a mocha test scheme for a working solution.
My reasoning for modifying the task was not as an attempt to solve the task, but rather to modify the scenario itself from a request for a terminal UI to a request for a simple balance tracker.

I think the most important thing to note about this fork is that im not trying to solve the tasks (and I don't believe that I have), I'm trying to modify the tasks into something that would be challenging to a new developer but not require many hours of their time either.

While yes, I was browsing uk.indeed and yes, I did link to my resume, I'm not currently looking for work. 
I find writing code entertaining and it's one of my favorite pass times. I know that companies often use tests like these, and I've been doing them as a sort of step up from algo/ds sites such as hackerrank. This one was quite different to some of the others, and while I did go ahead and solve them I felt like I didn't get the learning experience I was looking for.
One (admittedly opinionated) overhaul and short conversation later, I do feel like I've learned something. 

> Also your "solutions" are committed (unless they're only partial solutions, at which point I'd dare say there's too much in each of them ) : )

Those would be the 'complete' solutions, but to my own implementations of the tasks rather than the original scenarios. I had committed the changes locally but not realised they had also be pushed upstream, I'll be deleting the fork shortly so hopefully no 'copying of homework' has gone on.

Thank you very much for your time today.

**Pull request closed without merging**